cluster/kubernetes/sync.go

```diff
@@ -2,31 +2,308 @@ package kubernetes
 	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	rest "k8s.io/client-go/rest"
+
 	"github.com/weaveworks/flux"
 	"github.com/weaveworks/flux/cluster"
+	kresource "github.com/weaveworks/flux/cluster/kubernetes/resource"
+	"github.com/weaveworks/flux/policy"
+	"github.com/weaveworks/flux/resource"
+)
+
+const (
+	syncSetLabel       = kresource.PolicyPrefix + "sync-set"
+	checksumAnnotation = kresource.PolicyPrefix + "sync-checksum"
+
+	// The namespace to presume if something doesn't have one, and we
+	// haven't been told what to use as a fallback. This is what
+	// `kubectl` uses when there's no config setting the fallback
+	// namespace.
+	DefaultFallbackNamespace = "default"
 )
 
+// Sync takes a definition of what should be running in the cluster,
+// and attempts to make the cluster conform. An error return does not
+// necessarily indicate complete failure; some resources may succeed
+// in being synced, and some may fail (for example, they may be
+// malformed).
+func (c *Cluster) Sync(spec cluster.SyncSet) error {
+	logger := log.With(c.logger, "method", "Sync")
+
+	// Keep track of the checksum of each resource, so we can compare
+	// them during garbage collection.
+	checksums := map[string]string{}
+
+	// NB we get all resources, since we care about leaving unsynced,
+	// _ignored_ resources alone.
+	clusterResources, err := c.getResourcesBySelector("")
+	if err != nil {
+		return errors.Wrap(err, "collating resources in cluster for sync")
+	}
+
+	cs := makeChangeSet()
+	var errs cluster.SyncError
+	for _, res := range spec.Resources {
+		id := res.ResourceID().String()
+		// make a record of the checksum, whether we stage it to
+		// be applied or not, so that we don't delete it later.
+		csum := sha1.Sum(res.Bytes())
+		checkHex := hex.EncodeToString(csum[:])
+		checksums[id] = checkHex
+		if res.Policy().Has(policy.Ignore) {
+			logger.Log("info", "not applying resource; ignore annotation in file", "resource", res.ResourceID(), "source", res.Source())
+			continue
+		}
+		if cres, ok := clusterResources[id]; ok && cres.Policy().Has(policy.Ignore) {
+			logger.Log("info", "not applying resource; ignore annotation in cluster resource", "resource", cres.ResourceID())
+			continue
+		}
+		resBytes, err := applyMetadata(res, spec.Name, checkHex)
+		if err == nil {
+			cs.stage("apply", res.ResourceID(), res.Source(), resBytes)
+		} else {
+			errs = append(errs, cluster.ResourceError{ResourceID: res.ResourceID(), Source: res.Source(), Error: err})
+			break
+		}
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.muSyncErrors.RLock()
+	if applyErrs := c.applier.apply(logger, cs, c.syncErrors); len(applyErrs) > 0 {
+		errs = append(errs, applyErrs...)
+	}
+	c.muSyncErrors.RUnlock()
+
+	if c.GC {
+		orphanedResources := makeChangeSet()
+
+		clusterResources, err := c.getResourcesInSyncSet(spec.Name)
```

Comment:
I must be missing something but ... by only getting the resources in the syncset (which ultimately are the resources in the Git repo), don't we exclude the ones deleted in the git repo, preventing from garbage collecting them?

Reply:
Scratch that, that's not how `getResourcesInSyncSet` works